### ISSUE TITLE
`AttributeError: 'EnvVariable' object has no attribute '_cast'` in Heroku Build

### ISSUE THREAD

**Issue by johnnymetz** [DataDog/dd-trace-py#11286]
I'm using `ddtrace==2.15.1` in my Django app hosted on Heroku. I'm using the [Datadog Heroku Buildpack](https://docs.datadoghq.com/agent/basic_agent_usage/heroku/).

When attempting to build my app on Heroku I'm running into the following error:

```
$ python manage.py collectstatic --noinput
       Traceback (most recent call last):
         File "/app/.heroku/python/lib/python3.11/site-packages/django/utils/module_loading.py", line 30, in import_string
           return cached_import(module_path, class_name)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
         File "/app/.heroku/python/lib/python3.11/site-packages/django/utils/module_loading.py", line 15, in cached_import
           module = import_module(module_path)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^
         File "/app/.heroku/python/lib/python3.11/importlib/__init__.py", line 126, in import_module
           return _bootstrap._gcd_import(name[level:], package, level)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
         File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
         File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
         File "<frozen importlib._bootstrap>", line 1126, in _find_and_load_unlocked
         File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
         File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
         File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
         File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
         File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
         File "<frozen importlib._bootstrap_external>", line 940, in exec_module
         File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
         File "/tmp/build_62fa6f5a/api/auth/__init__.py", line 2, in <module>
           from api.auth.datadog import datadog_login
         File "/tmp/build_62fa6f5a/api/auth/datadog.py", line 1, in <module>
           from ddtrace import tracer
         File "/app/.heroku/python/lib/python3.11/site-packages/ddtrace/__init__.py", line 22, in <module>
           from ._monkey import patch  # noqa: E402
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
         File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
         File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
         File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
         File "/app/.heroku/python/lib/python3.11/site-packages/ddtrace/internal/module.py", line 295, in _exec_module
           self.loader.exec_module(module)
         File "/app/.heroku/python/lib/python3.11/site-packages/ddtrace/_monkey.py", line 12, in <module>
           from .settings.asm import config as asm_config
         File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
         File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
         File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
         File "/app/.heroku/python/lib/python3.11/site-packages/ddtrace/internal/module.py", line 295, in _exec_module
           self.loader.exec_module(module)
         File "/app/.heroku/python/lib/python3.11/site-packages/ddtrace/settings/asm.py", line 240, in <module>
           _report_telemetry(config)
         File "/app/.heroku/python/lib/python3.11/site-packages/ddtrace/settings/_core.py", line 20, in report_telemetry
           if env_name in env.source and env_val == e._cast(e.type, raw_val, env):
                                                    ^^^^^^^
       AttributeError: 'EnvVariable' object has no attribute '_cast'
       The above exception was the direct cause of the following exception:
       Traceback (most recent call last):
         File "/app/.heroku/python/lib/python3.11/site-packages/rest_framework/settings.py", line 179, in import_from_string
           return import_string(val)
                  ^^^^^^^^^^^^^^^^^^
         File "/app/.heroku/python/lib/python3.11/site-packages/django/utils/module_loading.py", line 32, in import_string
           raise ImportError(
       ImportError: Module "api.auth.rest_framework" does not define a "TokenAuthentication" attribute/class
       During handling of the above exception, another exception occurred:
       Traceback (most recent call last):
         File "/tmp/build_62fa6f5a/manage.py", line 21, in <module>
           main()
         File "/tmp/build_62fa6f5a/manage.py", line 17, in main
           execute_from_command_line(sys.argv)
         File "/app/.heroku/python/lib/python3.11/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
           utility.execute()
         File "/app/.heroku/python/lib/python3.11/site-packages/django/core/management/__init__.py", line 416, in execute
           django.setup()
         File "/app/.heroku/python/lib/python3.11/site-packages/django/__init__.py", line 24, in setup
           apps.populate(settings.INSTALLED_APPS)
         File "/app/.heroku/python/lib/python3.11/site-packages/django/apps/registry.py", line 124, in populate
           app_config.ready()
         File "/tmp/build_62fa6f5a/store/apps.py", line 9, in ready
           import store.signals
         File "/tmp/build_62fa6f5a/store/signals.py", line 9, in <module>
           from store.data_warehouse import DataWarehouseClient
         File "/tmp/build_62fa6f5a/store/data_warehouse/__init__.py", line 2, in <module>
           from .helpers import *
         File "/tmp/build_62fa6f5a/store/data_warehouse/helpers.py", line 7, in <module>
           from store import models, serializers
         File "/tmp/build_62fa6f5a/store/serializers.py", line 14, in <module>
           from api.serializers import defaults
         File "/tmp/build_62fa6f5a/api/serializers/__init__.py", line 7, in <module>
           from .deployments import *
         File "/tmp/build_62fa6f5a/api/serializers/deployments.py", line 18, in <module>
           from api.serializers.organization_invitation import OrganizationSlimSerializer
         File "/tmp/build_62fa6f5a/api/serializers/organization_invitation.py", line 3, in <module>
           from rest_framework.generics import get_object_or_404
         File "/app/.heroku/python/lib/python3.11/site-packages/rest_framework/generics.py", line 9, in <module>
           from rest_framework import mixins, views
         File "/app/.heroku/python/lib/python3.11/site-packages/rest_framework/views.py", line 17, in <module>
           from rest_framework.schemas import DefaultSchema
         File "/app/.heroku/python/lib/python3.11/site-packages/rest_framework/schemas/__init__.py", line 33, in <module>
           authentication_classes=api_settings.DEFAULT_AUTHENTICATION_CLASSES,
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
         File "/app/.heroku/python/lib/python3.11/site-packages/rest_framework/settings.py", line 227, in __getattr__
           val = perform_import(val, attr)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^
         File "/app/.heroku/python/lib/python3.11/site-packages/rest_framework/settings.py", line 170, in perform_import
           return [import_from_string(item, setting_name) for item in val]
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
         File "/app/.heroku/python/lib/python3.11/site-packages/rest_framework/settings.py", line 170, in <listcomp>
           return [import_from_string(item, setting_name) for item in val]
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
         File "/app/.heroku/python/lib/python3.11/site-packages/rest_framework/settings.py", line 182, in import_from_string
           raise ImportError(msg)
       ImportError: Could not import 'api.auth.rest_framework.TokenAuthentication' for API setting 'DEFAULT_AUTHENTICATION_CLASSES'. ImportError: Module "api.auth.rest_framework" does not define a "TokenAuthentication" attribute/class.
 !     Error: Unable to generate Django static files.
 !     
 !     The 'python manage.py collectstatic --noinput' Django
 !     management command to generate static files failed.
 !     
 !     See the traceback above for details.
 !     
 !     You may need to update application code to resolve this error.
 !     Or, you can disable collectstatic for this application:
 !     
 !        $ heroku config:set DISABLE_COLLECTSTATIC=1
 !     
 !     https://devcenter.heroku.com/articles/django-assets
 !     Push rejected, failed to compile Python app.
 !     Push failed
```

Note running this `collectstatic` command locally works fine.

Downgrading to `ddtrace=2.14.4` fixes the issue.

**Comment by ncvc** [comment]
I'm getting the same issue - quick workaround is to upgrade the `envier` library to at least 0.5.2.

ddtrace depends on `envier` versions `>=0.5,<1.0`. However, the `EnvVariable._cast` method was added in version [0.5.2](https://github.com/DataDog/envier/releases/tag/v0.5.2) [here](https://github.com/DataDog/envier/pull/35/files#diff-328fcc831dd05af9986a37698ee9efaf88254ada077ef48a999221e6fe0551a5R70).

I think the solution is for ddtrace to depend on `envier` versions `>=0.5.2,<1.0`. Though because `EnvVariable._cast` is not a public API, any upgrade to the envier library could break ddtrace, so I'd suggest removing usages of that method call (not familiar enough with ddtrace's code to know if that's feasible or not).

**Comment by wconti27** [comment]
@ncvc This workaround looks good to me.

Hey, please install the latest version of envier & ddtrace. The issue should be resolved. ddtrace==2.15.1 introduced an envier regression which was fixed in 2.16

**Comment by em-synth** [comment]
I am seeing this error on ddtrace 2.17.2. Is this really fixed on 2.16?

**Comment by evanr76** [comment]
Still happening in ddtrace 2.19.1.  Upgrading `envier` to 0.5.2 still fixes.

**Comment by adamsxs** [comment]
Ran into this as well with ddtrace==2.20.0 and on ddtrace==2.21.1. The problem is that the requirement for envier, `"envier~=0.5",` will be satisfied if you already have envier==0.5.1 installed or are using something like PDM and have envier==0.5.1 in your lock file. Upgrading envier fixes the problem, and I've found it to work with envier==0.6.1.
Setting the package requirement to `envier>=0.5.2` would be ideal instead.
